UPE Subscriptions support

ricardo opened the pull request:
Fixes #1665
Fixes #1785

# Changes proposed in this Pull Request:

- Refactor existing subscription compatibility classes (CC and SEPA) into a trait.
- Copy over [subscriptions utilities trait from WCPay](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/compat/subscriptions/trait-wc-payments-subscriptions-utilities.php).
- Make the subscriptions-compatible gateway classes use the subscriptions trait, and adapt subscriptions functions such as `save_source_to_order` and `process_payment` to be called from the gateway class, because traits don't have parent classes.
- Adapt `process_subscription_payment`, so the same trait can be used for card, UPE and SEPA payments.
- Some adaptations had to be made in order to support `sources` and `payment_methods` at the same time.
- Minor fixes to tests.

# Known issues

### Non-UPE subscriptions

Everything should work as expected, both for cards and SEPA. Except for a couple known issues that are currently being worked on:

| Description | Issue |
|:--|--|
| ~Pay for order page is not working. Trying to change a payment method or paying for a subscription from a guest order doesn't work.~  | https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1856  |
| ~SCA authorization flow is not working, therefore subscription + SCA cards is not working~ | https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1866 |

### UPE subscriptions

| Description | Issue |
|:--|--|
| 🚧  Free trial subscriptions not working. | https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1867 |
| ❌ Automatically renewing SEPA payments doesn't work. It seems like SEPA payment intents that are paid using saved payment methods (`pm_`) are not properly processed with webhooks. - This is probably an issue with webhooks or the SEPA implementation. | https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1850 |
| ~Renewing a subscription as the shopper, having `WooCommerce > Settings > Subscriptions > Accept Early Renewal Payments via a Modal` enabled, doesn't work.~ | TBD |
| ~Changing the payment method as the shopper triggers subscription renewal.~ |  |

# Testing instructions

To summarize, I'll reference the testing instructions for subscription flows. Please test these with and without UPE enabled.

### Shopper (subscriptions)

* [Purchase subscription product](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#purchase-subscription-product)
* [Purchase multiple subscriptions (same schedule)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#purchase-multiple-subscriptions-same-schedule)
* [Purchase multiple subscriptions (differing schedules)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#purchase-multiple-subscriptions-differing-schedules)
* [Renew subscription](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#renew-subscription)
* [Change payment method to new card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#change-payment-method-to-new-card)
* [Change payment method to saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#change-payment-method-to-saved-card)
* [Set default payment method](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#set-default-payment-method)
* [Change default payment method](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#change-default-payment-method)

### Merchant (subscriptions)

* [Renew subscription manually](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#renew-subscription-manually)
* [Renew subscription automatically](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#renew-subscription-automatically)

ricardo committed:
Refactor subscriptions compat class into a trait

ricardo committed:
Move subscriptions compatibility trait to abstract `WC_Stripe_Payment_Gateway`

ricardo committed:
Refactor early renewal actions on intent success or failure

ricardo committed:
Fix existing unit tests

ricardo committed:
Check if subscriptions is enabled before processing early renewals

ricardo committed:
Add subscriptions utilities trait

ricardo committed:
Use same trait for SEPA subscriptions

ricardo committed:
Fallback to charges API when SEPA is used

ricardo committed:
Save `pm_id` in `_stripe_source_id` metadata

ricardo committed:
Remove redundant minimum amount check

ricardo committed:
Fixing a couple misc leftovers and comments

ricardo committed:
Fix duplicate checkbox from payment methods page

ricardo committed:
Move `check_source` to abstract class

ricardo committed:
Force save payment method if order contains subscription

ricardo committed:
Fix payment method not showing up in the "View subscription" page

ricardo commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -263,7 +256,7 @@ public function payment_fields() {
 		}
 
-		do_action( 'wc_stripe_cards_payment_fields', $this->id );
+		do_action( 'wc_stripe_payment_fields_stripe', $this->id );
```
I changed the payment fields hooks for CC and SEPA, so it's easier to add the hook from the trait using `$this->id`.

ricardo commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -550,6 +531,14 @@ public function process_payment( $order_id, $retry = true, $force_save_source =
 		try {
 			$order = wc_get_order( $order_id );
 
+			if ( $this->has_subscription( $order_id ) ) {
+				$force_save_source = true;
+			}
+
+			if ( $this->maybe_change_subscription_payment_method( $order_id ) ) {
+				return $this->process_change_subscription_payment_method( $order_id );
+			}
```
This logic used to be in the subs compat class, but since we can't overwrite `process_payment`, we need to add it here.

Ideally we should add to just one place, but unfortunately all gateways have their own `process_payment` and don't call `parent::process_payment`. 🙁

ricardo commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -4,78 +4,62 @@
-	public function maybe_hide_save_checkbox( $display_tokenization ) {
-		if ( WC_Subscriptions_Cart::cart_contains_subscription() ) {
-			return false;
+	public function maybe_init_subscriptions() {
```
Instead of the old construct, this adds all the same hooks for cards and SEPA and overwrites the `supports` prop.

ricardo commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -231,29 +192,21 @@ public function scheduled_subscription_payment( $amount_to_charge, $renewal_orde
 	 */
 	public function process_subscription_payment( $amount, $renewal_order, $retry = true, $previous_error = false ) {
 		try {
-			if ( $amount * 100 < WC_Stripe_Helper::get_minimum_amount() ) {
-				/* translators: minimum amount */
-				$message = sprintf( __( 'Sorry, the minimum allowed order total is %1$s to use this payment method.', 'woocommerce-gateway-stripe' ), wc_price( WC_Stripe_Helper::get_minimum_amount() / 100 ) );
-				throw new WC_Stripe_Exception(
-					'Error while processing renewal order ' . $renewal_order->get_id() . ' : ' . $message,
-					$message
-				);
-			}
```
We don't need this because `validate_minimum_order_amount` does the same thing.

ricardo commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1629,24 +1640,6 @@ public function create_and_confirm_intent_for_off_session( $order, $prepared_sou
-	 *
-	 * @param int $order_id subscription renewal order id.
-	 */
-	public function ensure_subscription_has_customer_id( $order_id ) {
-		$subscriptions_ids = wcs_get_subscriptions_for_order( $order_id, [ 'order_type' => 'any' ] );
-		foreach ( $subscriptions_ids as $subscription_id => $subscription ) {
-			if ( ! metadata_exists( 'post', $subscription_id, '_stripe_customer_id' ) ) {
-				$stripe_customer = new WC_Stripe_Customer( $subscription->get_user_id() );
-				update_post_meta( $subscription_id, '_stripe_customer_id', $stripe_customer->get_id() );
-				update_post_meta( $order_id, '_stripe_customer_id', $stripe_customer->get_id() );
-			}
-		}
-	}
```
I have a note to test this more thoroughly, but apparently it doesn't make sense anymore.

ricardo commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -231,29 +192,21 @@ public function scheduled_subscription_payment( $amount_to_charge, $renewal_orde
-			if ( isset( $_REQUEST['process_early_renewal'] ) ) { // wpcs: csrf ok.
-				$response = parent::process_payment( $order_id, true, false, $previous_error, true );
+			if ( isset( $_REQUEST['process_early_renewal'] ) && 'stripe' === $this->id ) { // phpcs:ignore WordPress.Security.NonceVerification
+				// - TODO: Fix UPE early renewals via client modal.
```
I added this to the PR description, but I will create an issue.

ricardo commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -314,10 +267,20 @@ public function process_subscription_payment( $amount, $renewal_order, $retry =
-			$this->lock_order_payment( $renewal_order );
+			// If the payment gateway is SEPA, use the charges API.
+			// TODO: Remove when SEPA is migrated to payment intents.
+			if ( 'stripe_sepa' === $this->id ) {
+				$request            = $this->generate_payment_request( $renewal_order, $prepared_source );
+				$request['capture'] = 'true';
+				$request['amount']  = WC_Stripe_Helper::get_stripe_amount( $amount, $request['currency'] );
+				$response           = WC_Stripe_API::request( $request );
 
-			$response                   = $this->create_and_confirm_intent_for_off_session( $renewal_order, $prepared_source, $amount );
-			$is_authentication_required = $this->is_authentication_required_for_payment( $response );
+				$is_authentication_required = false;
```
Also added this to the PR description and need to create an issue, but to summarize:

SEPA payments _could_ be processed through payment intents, but apparently the webhooks are not updating SEPA `payment_intent.succeeded` for some reason.

It works for the charges API though, so I'm keeping this for BW compatibility.

ricardo commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -414,7 +381,12 @@ public function save_source_to_order( $order, $source ) {
 			update_post_meta( $subscription_id, '_stripe_customer_id', $source->customer );
-			update_post_meta( $subscription_id, '_stripe_source_id', $source->source );
+
+			if ( ! empty( $source->payment_method ) ) {
+				update_post_meta( $subscription_id, '_stripe_source_id', $source->payment_method );
+			} else {
+				update_post_meta( $subscription_id, '_stripe_source_id', $source->source );
+			}
```
[Historically](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/f63fb79d4d9058f0625d64ea9adda5ceec85c38b/includes/compat/trait-wc-stripe-subscriptions.php#L503), payment methods are also stored as `_stripe_source_id`, probably because we used to implement a `cards` API which saved a meta key which is now in `_stripe_source_id` too.

Another alternative would be to save it to a different key, but I didn't go with that because that seemed like a much bigger diff and because we're already expecting a `pm_` token in the `_stripe_source_id`.

ricardo commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -926,41 +948,6 @@ public function set_payment_method_title_for_order( $order, $payment_method_type
-	 * @param WC_Order         $order The order.
-	 * @param WC_Payment_Token $token The token to save.
-	 */
-	public function add_token_to_order( $order, $token ) {
```
Unfortunately Stripe doesn't use order tokens, it uses post meta, so I had to adapt the old code here instead.

ricardo reviewed (commented):
Added a few comments below ⬇️. Please let me know if anything is unclear.

ricardo committed:
Fix mistake with request body

FangedParakeet commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -1070,6 +1059,7 @@ public function verify_intent_after_checkout( $order ) {
 	 */
 	protected function handle_intent_verification_success( $order, $intent ) {
 		$this->process_response( end( $intent->charges->data ), $order );
+		$this->maybe_process_subscription_early_renewal_success( $order, $intent );
```
I couldn't find this function defined anywhere. Is there something I'm missing here or is this extended from core or something?

ricardo replied:
Good catch! That was my mistake. Fixed in 05e34d18f77f13ae109006b3d4541191e4b5d465

FangedParakeet commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -1081,6 +1071,7 @@ protected function handle_intent_verification_success( $order, $intent ) {
 	 */
 	protected function handle_intent_verification_failure( $order, $intent ) {
 		$this->failed_sca_auth( $order, $intent );
+		$this->maybe_process_subscription_early_renewal_failure( $order, $intent );
```
Same thing here--I couldn't find this function defined anywhere either. Where is this supposed to come from?

ricardo replied:
Same as the above. Fixed in 05e34d18f77f13ae109006b3d4541191e4b5d465

FangedParakeet commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -33,6 +33,13 @@ class WC_Stripe_UPE_Payment_Gateway extends WC_Stripe_Payment_Gateway {
+	 * @var int
+	 */
+	public $retry_interval;
+
```
```suggestion
```
FWIW, this is already declared on `WC_Stripe_Payment_Gateway` over [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/6e8741630fd1c3e4f0cf7bcf8685891e05f40e9d/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L21), which this gateway extends.

ricardo replied:
Oh, you've added it 😅. I rebased and didn't notice. Fixed in aadd58adf289a277509545cca1fcab2f092da2b3

FangedParakeet commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -79,8 +86,9 @@ class WC_Stripe_UPE_Payment_Gateway extends WC_Stripe_Payment_Gateway {
-		$this->id           = self::ID;
-		$this->method_title = __( 'Stripe UPE', 'woocommerce-gateway-stripe' );
+		$this->id             = self::ID;
+		$this->retry_interval = 1;
```
```suggestion
```
This is redundant, as it's declared and defined on the parent gateway over [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/6e8741630fd1c3e4f0cf7bcf8685891e05f40e9d/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L21).

ricardo replied:
Also fixed in aadd58adf289a277509545cca1fcab2f092da2b3.

FangedParakeet commented on includes/compat/trait-wc-stripe-subscriptions.php:
```diff
@@ -146,14 +131,31 @@ public function differentiate_change_payment_method_form() {
+	public function process_change_subscription_payment_method( $order_id ) {
 		try {
 			$subscription    = wc_get_order( $order_id );
 			$prepared_source = $this->prepare_source( get_current_user_id(), true );
```
When changing subscription payment method to another payment method that has been saved via the UPE `prepare_source` will throw an Exception ("No such source: 'pm_...'"), because `prepare_source` (I think [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L818)) will attempt to retrieve a source using a saved payment method ID.

ricardo replied:
Good callout! This was something I had a note about but forgot to fix. Finally fixed in d3787aa4c87926a2f9080195d9a9d9da57a53ef2.

FangedParakeet reviewed (changes requested):
I tested this using the non-UPE card and SEPA gateways, as well as all the reusable payment methods in the UPE, following the critical testing flows listed out [here](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows). I was able to do some things, but unable to do other things: I'm not entirely sure what the full scope of this PR is supposed to be, so I'll just list out what worked and what didn't and you can let me know if it's supposed to work here or elsewhere.

### Cards (Non-UPE)

I was able to complete all the purchasing subscription flows and everything worked as expected. I was also able to renew a subscription. 

I was able to change my payment method to another existing saved payment method; however, if I tried to change the payment method on a subscription to a new payment method, I received the error "Your payment information is incomplete" on a loop, as it appeared as if the page was retrying and failing to process this request over and over infinitely.

### SEPA (Non-UPE)

I was not able to purchase any subscription product with the SEPA gateway, sadly. During the checkout process, the SEPA payment method was saved, but eventually I received the standard "Sorry, we are unable to process your payment at this time. Please retry later." error message. My subscription order was created, but left with a status of `Pending`.

### UPE 

I was able to complete all the subscription purchase flows (mostly), using all payment methods. I was also able to renew a subscription.

However, for some reason, the first time I tried to purchase a subscription for an account, I was redirected to a page titled "Something went wrong." containing the message "The link you followed has expired." I could repeat this by logging out of the customer account I was signed into, adding a subscription product to cart, and simply trying to complete a checkout using any UPE payment method. After seeing this page, I could return to the checkout, finding the payment method I used saved, and could then successfully complete the checkout.

![Weird error page](https://cdn-std.droplr.net/files/acc_1104206/P4QAJv)
_Weird error page_

I encountered a few more issues while trying to change my payment method. If I tried to change my payment method to a new payment method, the form submitted and I returned back at the main account page with no success message; upon reviewing the subscription, I could see that it was still using the original payment method instead of the new one. If I tried to change my payment method to an existing saved payment method, I received the error "No such source 'pm_...'". I left a comment below that might explain why this error occurred.

It's also worth noting that on the change payment method page, the available payment methods were not filtered by their ability to be used for recurring payments, e.g. both giropay and EPS were available as selectable payment methods, despite neither being supported for subscriptions. Also, I was only able to add card payment methods from this screen. For anything else, I think the form was expecting the regular checkout billing fields to be there, so I received errors like the below. I'm not sure if this is related to other issues on the Pay for Order page, since #1785 also exists.

> When confirming a PaymentIntent with a `setup_future_usage` value of `off_session` and a PaymentMethod of type `ideal`, the PaymentMethod must include `billing_details[name]` and `billing_details[email]`. Please provide those details and try confirming again.

Let me know your thoughts on the above (and below) and let me know if my testing has exceeded the intended scope of this PR. Cheers!

ricardo committed:
Fix method name in `WC_Stripe_Subscriptions_Trait`

ricardo committed:
Remove redundant `retry_interval` from `WC_Stripe_UPE_Payment_Gateway`

ricardo committed:
Fix unit tests and SEPA regression

ricardo commented:
@FangedParakeet thanks for reviewing it so quickly and for the detailed test report!

I added known issues to the PR description, but I should have more clearly described what wasn't expected to work at this point.

> I received the error "Your payment information is incomplete" on a loop, as it appeared as if the page was retrying and failing to process this request over and over infinitely.

That was part of #1856. It's fixed now that I've rebased.

> I was not able to purchase any subscription product with the SEPA gateway, sadly.

There was a regression from my own commits :grimacing:. Fixed it in bce6de66ba97ab6baec53ff12e62d9cfac8675de. The error "When confirming a PaymentIntent with a `setup_future_usage` value of `off_session`..." was also part of this.

But the orders shouldn't be going to "Pending". Please ensure you are listening to webhooks.

```ssh
stripe login
stripe listen --forward-to 'http://localhost:8082/?wc-api=wc_stripe'
```

The below issues were indeed problems I didn't foresee. I'm working on them right now:
- > I was redirected to a page titled "Something went wrong." containing the message "The link you followed has expired."
- > I received the error "No such source 'pm_...'"
- > the available payment methods were not filtered by their ability to be used for recurring payments

FangedParakeet commented:
Nice one. I think a lot of my troubles started on the Pay for Order page, so if that's now resolved in #1856 that hopefully takes care of a lot of the oddities I described. I'm also pretty sure I have not been listening to webhooks properly, so that will probably account for a few more. Just gimme a ping when you're ready for me to have another look and I'll run back the tape and let you know how I get on!

ricardo committed:
Fix error "No such source 'pm_...'"

ricardo committed:
Check if order contains subscription on `create_payment_intent` AJAX and display payment method accordingly. This is necessary for renewal and changing payment method for an existing order.

ricardo committed:
Disable iDeal and Bancontact for off session payments as they're both not working with the "Add payment method" page yet

ricardo committed:
Move prepaid card check to abstract, so it can be used by SEPA as well

ricardo committed:
Specify `sepa_debit` in `payment_method_types` when it's enabled for payment intent requests

ricardo commented:
> I was redirected to a page titled "Something went wrong." containing the message "The link you followed has expired."

I couldn't reproduce this issue. Can you try again to see if it's reproducible by the steps you mentioned? You shouldn't be able to checkout without being authenticated.

> the available payment methods were not filtered by their ability to be used for recurring payments

There was indeed a bug which was fixed with 013b6ae88e4fa219b7a006a3d0662c2b1a2ad29f, but when renewing a subscription early you should be able to see the other payment methods.

I have disabled EPS and Bancontact because they don't seem to be working as off session methods and don't work for the "Add payment method" page.

Subscriptions and pre-orders weren't supported for these methods and they're not documented as reusable in the plugin like SEPA and credit card under `WooCommerce > Settings > Payments > Automatic Recurring Payments`, so I guess that should be ok.

---

@FangedParakeet, thanks for your patience 😅. I think you can give it another review, and then we can fix the remainder issues once other blocking PRs are merged (#1850, #1867 and the modal thing).

I've documented below what I've tested more recently. I don't expect you to go through all these cases (some of which are kinda redundant), so feel free to stick to the testing instructions from the critical flows. GlobalStep should also give us some more confidence once they finish their business. I'll make sure to add extra testing instructions for this release in case this PR gets to the UPE MVP.

Note: When testing SEPA via Stripe CLI, often times people have issues with receiving the events. I sometimes don't get the `charges.succeeded` event, so I have to restart the CLI and try again.

## non-UPE

#### Credit card

1. ✅ Regular checkout subscription with new CC.
2. ✅ Regular checkout subscription with saved CC.
3. ✅ Renew subscription as merchant.
4. ✅ Renew subscription as shopper with new CC.
5. ✅ Renew subscription as shopper with existing CC.
6. ✅ Change payment method to new CC.
7. ✅ Change payment method to saved CC.
8. ✅ Renew subscription as merchant after payment method change.
9. ✅ Add new payment method and set it to default.
10. ✅ Remove other CC and renew subscription using default.
11. ✅ Renew subscription as shopper with modal (`WooCommerce > Settings > Subscriptions > Accept Early Renewal Payments via a Modal`).
12. ✅ Buy subscription with SCA card that requires authentication for on session payments (`4000002500003155`).
13. ✅ Renew subscription as merchant with SCA card that requires auth for on session payments.
14. ✅ Renew subscription as shopper with SCA card that requires auth for on session payments.
15. ✅ Buy subscription with SCA card that always requires authentication (`4000002760003184`).
16. ✅ Renew subscription as merchant with SCA card that always requires authentication. - It shouldn't work.
17. ✅ Renew subscription as shopper with SCA card that always requires authentication.


#### SEPA

1. ✅ Regular checkout subscription with new SEPA.
2. ✅ Regular checkout subscription with existing SEPA.
3. ✅ Renew subscription as merchant.
4. ✅ Renew subscription as shopper with new SEPA.
5. ✅ Renew subscription as shopper with existing SEPA.
6. ✅ Change payment method to new SEPA.
7. ✅ Change payment method to existing SEPA.
8. ✅ Renew subscription as merchant after payment method change.
9. ✅ Add new payment method and set it to default.
10. ✅ Remove other payment methods and renew subscription as merchant using default.
11. ✅ Renew subscription as shopper with modal (`WooCommerce > Settings > Subscriptions > Accept Early Renewal Payments via a Modal`).

## UPE

#### Credit card

1. ✅ Regular checkout subscription with new CC.
2. ✅ Regular checkout subscription with saved CC.
3. ✅ Renew subscription as merchant.
4. ✅ Renew subscription as shopper with new CC.
5. ✅ Renew subscription as shopper with existing CC.
6. ✅ Change payment method to new CC.
7. ✅ Change payment method to saved CC.
8. ✅ Renew subscription as merchant after payment method change.
9. ✅ Add new payment method and set it to default.
10. ✅ Remove other CC and renew subscription using default.
11. ❌  Renew subscription as shopper with modal (`WooCommerce > Settings > Subscriptions > Accept Early Renewal Payments via a Modal`).
12. ✅ Buy subscription with SCA card that requires authentication for on session payments (`4000002500003155`).
13. ✅ Renew subscription as merchant with SCA card that requires auth for on session payments.
14. ✅ Renew subscription as shopper with SCA card that requires auth for on session payments.
15. ✅ Buy subscription with SCA card that always requires authentication (`4000002760003184`).
16. ✅ Renew subscription as merchant with SCA card that always requires authentication. - It shouldn't work.
17. ✅ Renew subscription as shopper with SCA card that always requires authentication.

#### SEPA
❌ Automatically renewing SEPA payments doesn't work due to not being able to process payment intent "succeeded" webhooks #1850 - therefore it's hard to test any SEPA UPE functionalities.

ricardo committed:
Fix subscription payment method change leading to order details page

ricardo committed:
Disable early renewal modal for UPE 3DS cards.

ricardo commented:
@FangedParakeet I've "fixed" the following issue by disabling the "renew early" modal for UPE SCA cards. There's no real negative impact in disabling that as the shopper will be redirected to the checkout page in case the card is a 3D Secure card.

- ✅ Renew subscription as shopper with modal (`WooCommerce > Settings > Subscriptions > Accept Early Renewal Payments via a Modal`).

I have another PR for trial subscriptions, so I think the only blocker issue for this should be the one related to SEPA statuses #1850.

PS: Closed the PR by mistake while writing this comment.

FangedParakeet commented:
@ricardo, I'm testing this now and still having a few issues with the non-UPE SEPA gateway, but let me know if you think they might be self-inflicted or related to any other issues.

I still get `Sorry, we are unable to process your payment at this time. Please retry later.` when checking out a subscription product using the SEPA gateway. Checking my error log, I can see a lot of `Incoming webhook failed validation` errors. Is this fixed by #1927 perhaps? Let me know if you have any ideas what's going on here, while I'll keep testing the rest...

P.S. I have my webhook listeners on this time. I can see `setup_intent.created`, among other things, but no `charge.succeeded` for SEPA transactions.

ricardo commented:
> `Incoming webhook failed validation` errors

@FangedParakeet That's probably due to the "Test Webhook Secret" under **WooCommerce > Settings > Stripe** not being properly set. When you run `stripe listen` you should get a webhook secret (`whsec_`). Paste that in the "Test Webhook Secret" input and that should work.

Let me know if you still have issues with it.

ricardo commented:
> I have my webhook listeners on this time. I can see setup_intent.created, among other things, but no charge.succeeded for SEPA transactions.

Also, it's common to not receive certain webhook events, specially `charge.succeeded`. This doesn't seem like an issue when you have a webhook URL configured in Stripe, so this leads us to think the `listen` command on the CLI is flawed as other people often have the same issues.

To fix this, I restart `stripe listen`. It works for a while and then often starts missing events again. I keep an eye for the events received to make sure they're being processed by the plugin.

FangedParakeet commented:
Argh, I had entered the right webhooks secret, but I kept overwriting it with the wrong one because I was manually updating `woocommerce_stripe_settings` in my `wp_options` table. 🤦‍♂️ D'oh.

Anyways, now I have the correct webhooks secret in there and I'm getting a new error with SEPA! Here is the error response from Stripe this time.

```
Error: stdClass Object
(
    [error] => stdClass Object
        (
            [code] => parameter_unknown
            [doc_url] => https://stripe.com/docs/error-codes/parameter-unknown
            [message] => Received unknown parameter: setup_future_usage
            [param] => setup_future_usage
            [type] => invalid_request_error
        )

)
```

For reference, here is the outgoing charges request from my logs as well.

```
charges request: Array
(
    [currency] => eur
    [amount] => 7320
    [description] => WooCommerce Stripe Dev - Order 238
    [statement_descriptor] => FANGED.PARAKEET
    [shipping] => Array
        (
            [name] => Mister Popo
            [address] => Array
                (
                    [line1] => 123 Fake Street
                    [line2] => 
                    [city] => 12345
                    [country] => US
                    [postal_code] => 12345
                    [state] => CA
                )

        )

    [expand[]] => balance_transaction
    [setup_future_usage] => off_session
    [metadata] => Array
        (
            [customer_name] => Mister Popo
            [customer_email] => mister@popo.com
            [order_id] => 238
            [site_url] => https://fangedparakeet.jurassic.tube
            [payment_type] => recurring
        )

    [customer] => cus_KH8n3kuiArZFzq
    [source] => src_1JcartJ5cIRIG92xgZg834Gp
)
```

Any ideas what's going wrong here?

ricardo commented:
@FangedParakeet That should be fixed previously. I had force-pushed this branch, perhaps the `pull` failed? Can you make sure you have the right version of this branch with `git reset --hard origin/fix/1665-upe-subscriptions-support`

FangedParakeet commented:
>  I was redirected to a page titled "Something went wrong." containing the message "The link you followed has expired."

Also, with the UPE gateway, I am still able to repeat this error every time.

> You shouldn't be able to checkout without being authenticated.

This could be the reason though. I am repeating this error by logging out of my customer account, then attempting to complete a checkout for a subscription without being logged into a customer account. During the checkout, a new user account is made for me using my billing details and my payment information is saved as usual, so it looks like the checkout is functioning as expected. However, I then get redirected to that "Something went wrong." page, instead of the order complete page, presumably. If I refresh the page, I am returned to the checkout and I can checkout successfully. Is that expected behaviour for the checkout if I'm not logged in? It sounds a bit weird to create an account for me during checkout and then send me to that error page anyways.

Let me know if you are unable to repeat that using the steps I described. 

By the way, everything else other than these two issues is now working wonderfully for me, so there is more good news than bad! 😅

FangedParakeet commented:
> git reset --hard origin/fix/1665-upe-subscriptions-support

Woohoo! This fixed it! SEPA gateway is working as expected now. Sorry about that, not sure what my local was looking at. Just let me know your thoughts about the "Something went wrong." thing above. If that's expected behaviour, because I'm not logged in, I don't have any other reason not to approve this!

FangedParakeet commented on includes/payment-methods/class-wc-stripe-upe-payment-method-bancontact.php:
```diff
@@ -19,7 +19,7 @@ public function __construct() {
 		$this->stripe_id            = self::STRIPE_ID;
 		$this->title                = 'Pay with Bancontact';
-		$this->is_reusable          = true;
+		$this->is_reusable          = false;
```
Is there any reason why you set `is_reusable` for this payment method and the iDEAL payment method to `false`? These payment methods do support recurring payments. Is this related to anything in #1929 perchance?

ricardo replied:
Yeah, I explained above, but in the middle of a giant comment.

> I have disabled ~EPS~ iDeal and Bancontact because they don't seem to be working as off session methods and don't work for the "Add payment method" page.

> Subscriptions and pre-orders weren't supported for these methods and they're not documented as reusable in the plugin like SEPA and credit card under WooCommerce > Settings > Payments > Automatic Recurring Payments, so I guess that should be ok.

In addition to those issues, Stripe documents it as a [single use payment method](https://stripe.com/docs/payments/bancontact/accept-a-payment#:~:text=Bancontact%20is%20a%20single%20use,the%20payment%20succeeded%20or%20failed.). Although it apparently can be saved as SEPA.

ricardo replied:
I can try to make it work with subscriptions if the other issues are fixed with #1929, but I guess supporting tokenization for these methods is not a priority as we didn't offer that possibility before UPE.

FangedParakeet replied:
OK, cool. Yeah, Bancontact get saved and converted into reusable SEPA payment methods. That's been implemented and is working from the checkout. I explain why I'm pretty sure it's working at checkout and not from the Add payment method page in [this comment](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1929#issuecomment-925145919), if you're intrigued. 

For what it's worth, I don't mind switching them off here; we can turn them back on in #1929 once that wobbly functionality is resolved.

ricardo commented:
> then attempting to complete a checkout for a subscription without being logged into a customer account.

@FangedParakeet do you mean you're still logged in an admin account or you're logged out of all accounts?

It should not be possible to checkout with a subscription without being logged in. Unless you have an old version of the subscriptions plugin?

How does your **WooCommerce > Settings > Accounts & Privacy** screen look like?

<img width="1069" alt="WooCommerce Accounts & Privacy page with subscriptions note highlighted" src="https://user-images.githubusercontent.com/5509901/134412937-1e7b8266-fbae-4e6c-8b9a-9568dbb682ff.png">

When I have a subscription product added to the cart and then try to checkout, I get the following:

<img width="476" alt="WooCommerce checkout page without a subscription and without being authenticated" src="https://user-images.githubusercontent.com/5509901/134413082-d6007cf5-ad91-4e54-82ee-440c1456c043.png">

FangedParakeet commented:
OK, I think I am able to checkout while logged out, because I have "Allow subscription customers to create an account during checkout" checked. 

![Create account during checkout](https://cdn-std.droplr.net/files/acc_1104206/jh2hVd)
_Create account during checkout_

This sounds like it could be an issue with subscriptions. Either way it seems somewhat beyond the scope of the work here and at worst merits a new issue being created to resolve it, IMO.

FangedParakeet reviewed (approved):
I tested all the flows I had done previously. All working happily now, with the exception of trying to checkout while logged out. If desired, we can create a new issue to resolve this, but I see no reason to halt the glorious march of progress here right now. LGTM: approved!

ricardo commented:
@FangedParakeet Thanks for the review!

I've created the issue https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1944, which I'll fix right after. It affects other products as well, not only subscriptions, but it depends on the Accounts & Privacy settings, so it's a bit unrelated.

I'll rebase the PR, smoke test the flows again and merge.

ricardo committed:
Update meta key with `_stripe_customer_id` in UPE tests
